### PR TITLE
Media: Do not trim all instances of video

### LINKED ISSUE #10409: Media: Video Trimming affects all instances of video
## Bug Description

This is a follow-up to https://github.com/GoogleForCreators/web-stories-wp/pull/9945#issuecomment-1007467673

When trimming a video within a story, currently all instances of that video within the story are being trimmed.

The `useProcessMedia` hook uses `updateElementsByResourceId` to update all instances of a video item within the story.

This makes sense for optimization and probably also muting. But for video trimming this is confusing, as it's not really expected behavior as a user.

## Expected Behaviour

When trimming one instance of a video, it should only trim the current instance, not all instances of it.

We should not use `updateElementsByResourceId` for trimming.

## Steps to Reproduce

1. Drag video from media library to a page
2. Drag the same video from media library to a page (same page or new one)
3. Trim first video
4. Observe that both are now trimmed once completed

## Additional Context

- Plugin Version: 1.16.0
- WordPress Version: 5.9
- Operating System: macOS
- Browser: Chrome

### PR DESCRIPTION
**Pull request by timarney**
## Context

When trimming a video within a story, currently all instances of that video within the story are being trimmed.

See: https://github.com/GoogleForCreators/web-stories-wp/issues/10409

## Summary

The existing code uses `canvasResourceId` & `updateExistingElements(canvasResourceId,  ...)` to updating existing elements.

See: https://github.com/GoogleForCreators/web-stories-wp/pull/10554/files#diff-284f5d5d8d1215d90a8fa6d991491806595892db0619411fbbea3b629868d366R66

When adding multiple copies of the same video element the canvasResourceId ends up getting duplicated "not unique"

When calling updateExistingElements with the canvasResourceId --- any elements with that ID will get updated

| Element # 1  |      Element # 2     | 
|----------|:-------------:|
| 153723757-df3bedd4-41c8-4a58-85b5-11a834e552c2 |  153723775-d4c56683-ea13-47d7-b8f3-a2788db846e7 |

## Relevant Technical Choices

This PR updates the trimming functions to update existing elements by element ID (which should be unique).

## To-do

Write tests as needed

## User-facing changes

Bug fix for existing trim functionality 

## Testing Instructions

This PR can be tested by following these steps:
1) Drag video from media library to a page
2)  Drag the same video from media library to a page (same page or new one)
3) Trim first video
4) Observe that only the item that selected to be trimmed was updated

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

Fixes #10409

### CODE REVIEW

File: includes/Media/Video/Trimming.php [hunk 1]
```diff
@@ -74,18 +74,22 @@ public function register_meta() {
...
+							'start'     => [
 								'description' => __( 'Start time.', 'web-stories' ),
 								'type'        => 'string',
 							],
-							'end'      => [
+							'end'       => [
 								'description' => __( 'End time.', 'web-stories' ),
 								'type'        => 'string',
 							],
+							'elementId' => [
```
Comment (timarney) [inline includes/Media/Video/Trimming.php]: Was getting a REST error --- will see if this can be removed

Comment (swissspidy) [reply includes/Media/Video/Trimming.php]: This definitely needs to be removed.

You're getting the warnings because you are accidentally passing `elementId` to the underlying REST API call.

This happens because you added `elementId` to `trimData` in `trimExistingVideo()`. That needs to be removed.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 2]
```diff
@@ -42,9 +42,14 @@ function useProcessMedia({
     updateElementsByResourceId: state.actions.updateElementsByResourceId,
   }));
 
+  const { updateElementById } = useStory((state) => ({
```
Comment (swissspidy) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js]: Can just merge this with the `useStory()` call above.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 3]
```diff
@@ -178,38 +220,42 @@ function useProcessMedia({
...
+          oldResource: oldCanvasResource,
+          resource,
+        });
         postProcessingResource(resource);
       };
 
       const onUploadProgress = ({ resource }) => {
-        const newResourceWithCanvasId = { ...resource, id: canvasResourceId };
-        updateExistingElements(resourceId, {
+        const newResourceWithCanvasId = { ...resource, id: elementId };
```
Comment (swissspidy) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js]: This would override the resource ID with the element ID. We don't want that. Instead, we want to preserve the original resource ID.

Comment (timarney) [reply packages/story-editor/src/app/media/utils/useProcessMedia.js]: Noting: We have a reference to the `original` resourceId under `original`

```
const trimData = {
        original: resourceId,
        start,
        end,
        elementId, // I might be able to remove this completely
      };
```

File: includes/Media/Video/Trimming.php (hunk 1 above)
Comment (timarney) [reply includes/Media/Video/Trimming.php]: Noting this was added to the trim data --- to keep a temporary reference (while trimming) to the item selected to be trimmed  ... after trimming it should be removed.

If there's a better way to hold this reference that would be ideal.

Comment (timarney) [reply includes/Media/Video/Trimming.php]: Given most of the code in the PR has been updated to use updateElementById vs updating by Resource ID we might not need to hold the elementId in the trim data at all.

**Comment by spacedmonkey** [comment]
Storing the element id isn't going to work here. What if the same trimmed video is used in multiple place in one story? Or in multiple stories. This fix doesn't feel workabout.

**Comment by timarney** [comment]
@spacedmonkey 

> Storing the element id isn't going to work here. 

The code as is isn't finished yet --- what I'm trying to accomplish here is keeping a reference the element that the user is intending to be trimmed 👉 but only while the element is being trimmed 👈 .   You would have more context around holding a longer reference.

With the current code on main multiple elements can end up with the same CanvasResourceID --- when a trim is requested and using the CanvasResourceID as the item to be trimmed they all get trimmed.

I did a test setting a window variable to hold the reference to the the item to be trimmed see https://github.com/GoogleForCreators/web-stories-wp/pull/10555.
```
// for demo purposes only --- this will not remain in the code
      if (window.TEMP_TRIM_ID) {
        if (element.id == window.TEMP_TRIM_ID && element.resource?.id === id) {
          console.log(`found trim item ${element.id}`, element);
```

===

This PR is trying to go down down that road (and the trimming part works) --- the larger change here is `updateElementById` potentially I don't need to hold the elementID (temp variable) in the trim data at all.

**Comment by swissspidy** [comment]
> Storing the element id isn't going to work here. What if the same trimmed video is used in multiple place in one story? Or in multiple stories. This fix doesn't feel workabout.

Not sure I follow you here 🤔 

We don't care about the same video being used in multiple stories. And we also don't care about the same trimmed video being used multiple times _in the same story_. The whole point of this ticket is to not care about it.

File: includes/Media/Video/Trimming.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Media/Video/Trimming.php]: Why does this need to saved at all. Couldn't we just delete it before submitted to api?

```
delete trimData.elementId;
```

Comment (swissspidy) [reply includes/Media/Video/Trimming.php]: It doesn't need to be saved, so yes it obviously needs to be removed at some point.

I think we are all on the same page that this is just transient data.

Comment (timarney) [reply includes/Media/Video/Trimming.php]: > Why does this need to saved at all. Couldn't we just delete it before submitted to api?

yes for sure 

> I think we are all on the same page that this is just transient data.

yes

Comment (spacedmonkey) [reply includes/Media/Video/Trimming.php]: You can delete it here

https://github.com/GoogleForCreators/web-stories-wp/blob/27c547c73de98ae361f340cc025122ece0441b9f/packages/story-editor/src/app/media/utils/useMediaUploadQueue/index.js#L247

File: packages/story-editor/src/components/videoTrim/provider.js [hunk 4]
```diff
@@ -56,15 +56,15 @@ function VideoTrimProvider({ children }) {
       return;
     }
     const lengthInSeconds = Math.round(endOffset / 1000 - startOffset / 1000);
+
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/videoTrim/provider.js]: New line?

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 5]
```diff
@@ -248,15 +293,15 @@ function useProcessMedia({
             ...oldResourceWithoutId,
             trimData,
           },
-          originalResourceId: canvasResourceId,
+          originalResourceId: elementId,
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js]: This now seems wrong. You are using a element id in place of a resource id.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js (hunk 3 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/media/utils/useProcessMedia.js]: Why are we using the element id in place of a resource id here. This is really confusing to me.

File: packages/story-editor/src/components/videoTrim/provider.js [hunk 6]
```diff
@@ -56,15 +56,15 @@ function VideoTrimProvider({ children }) {
...
     trimExistingVideo({
       resource: {
         ...resource,
         length: lengthInSeconds,
         lengthFormatted: getVideoLengthDisplay(lengthInSeconds),
       },
       // This is the ID of the resource that's currently on canvas and needs to be cloned.
-      // It's only different from the above resource, if the canvas resource is a trim of the other.
-      canvasResourceId: element.resource.id,
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/videoTrim/provider.js:67]: Why does this need to be removed?

File: packages/story-editor/src/components/videoTrim/provider.js (hunk 4 above)
Comment (timarney) [reply packages/story-editor/src/components/videoTrim/provider.js]: removed

File: packages/story-editor/src/components/videoTrim/provider.js (hunk 6 above)
Comment (timarney) [reply packages/story-editor/src/components/videoTrim/provider.js:67]: As far as I can see it's not used anywhere else.  Correct me if I'm wrong but it's only used as a parameter for calling `trimExistingVideo `

Which I updated to 
```
const trimExistingVideo = useCallback(
    ({ resource: oldResource, elementId, start, end }) => {
 ```

File: packages/story-editor/src/app/media/utils/useProcessMedia.js (hunk 5 above)
Comment (timarney) [reply packages/story-editor/src/app/media/utils/useProcessMedia.js]: This one was going to ask you about and I want to dig a little further to understand that that is doing.

We have a reference to the `"original_id: resourceId"` under `"additionalData"` which is passed to that function.   So is "originalResourceId" just a temporary reference ?

```
 additionalData: {
            original_id: resourceId,
            web_stories_is_muted: isMuted,
            web_stories_media_source: isOptimized
              ? 'video-optimization'
              : 'editor',
          },
```

File: packages/story-editor/src/components/videoTrim/provider.js (hunk 6 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/videoTrim/provider.js:67]: It is used here https://github.com/GoogleForCreators/web-stories-wp/blob/e00e50da7a02d956b993f4038ca1c15d46eda3ef/packages/story-editor/src/app/media/utils/useProcessMedia.js#L211

I don't love the idea of elemnt and resource having the same id. It doesn't make it a unique identifier anymore and makes this code confusing and hard to follow.

Comment (swissspidy) [reply packages/story-editor/src/components/videoTrim/provider.js:67]: That ID could theoretically be derived from the element.

As discussed yesterday I am going through this PR in detail. Haven't gotten around to do that yet, but it's high up in my todo list.

Let's hold off on reviewing anything until then.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js (hunk 5 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/media/utils/useProcessMedia.js]: Why was this changed? Why does `originalResourceId` have to be the elementId?

**Comment by swissspidy** [comment]
@timarney Just had an initial look at this. As @spacedmonkey already pointed out, the resource ID should not be overridden with the element ID, so I made that change in 298ee26614e8951a5745f4ae8e46f3cb904ff090.

Now this already works quite well.

A couple of minor nits in this PR:

* Adding `elementId` to `trimData` doesn't feel great, but I can't think of a better place right now without adding new fields everywhere in the media upload queue. So I suppose we can live with it.
* The naming of all these callbacks (`copyResourceData` / `copyResourceDataByElementId` / `updateExistingElement` / `updateExistingElements`) is a bit confusing, maybe we can find better names

Also a general observation:

When adding the same video multiple times to the canvas and trimming only one of them, each video's "Trim" button changes to "Trimming..." and gets disabled, because technically the underlying resource actually is being trimmed.

However, trimming one video doesn't really affect the others, so their buttons shouldn't get disabled in that case.

Not sure what's a good way to address that, but it feels like it can be its own ticket as it could be a bigger lift. @spacedmonkey Would appreciate your thoughts on that.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 7]
```diff
@@ -38,13 +38,17 @@ function useProcessMedia({
...
+    (state) => ({
+      updateElementsByResourceId: state.actions.updateElementsByResourceId,
+      updateElementById: state.actions.updateElementById,
+    })
+  );
 
   const copyResourceData = useCallback(
     ({ oldResource, resource }) => {
       const { id, alt } = oldResource;
+
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js:51]:
```suggestion
```

**Comment by spacedmonkey** [comment]
> The naming of all these callbacks (copyResourceData / copyResourceDataByElementId / updateExistingElement / updateExistingElements) is a bit confusing, maybe we can find better names

Naming here is important to make this code readable. 

> When adding the same video multiple times to the canvas and trimming only one of them, each video's "Trim" button changes to "Trimming..." and gets disabled, because technically the underlying resource actually is being trimmed.

Yeah, I remember discussing this at the time. It is extremely unlikely the a user would be trimming the same video at the time. But the implementation blocks that by design. Consider the user story. 

1. User insides video. 
2. User trims video. Trim starts 
3. User insides video again. 
4. User tries to trim video, functionality is disabled. 

Now that update is based on element id and resource id, it does open the door to allow for multiple trimming of the same video. I do think this should be done in a follow up ticket.

**Review by spacedmonkey — APPROVED** [review]
Approved with one nic pic.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js (hunk 7 above)
Comment (swissspidy) [reply packages/story-editor/src/app/media/utils/useProcessMedia.js:51]: What's wrong with some extra breathing room?

**Comment by swissspidy** [comment]
> Naming here is important to make this code readable.

Any ideas for better names?

> Now that update is based on element id and resource id, it does open the door to allow for multiple trimming of the same video. I do think this should be done in a follow up ticket.

I'll create the ticket, but I am curious to hear some initial thoughts on how it could be fixed.

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 8]
```diff
@@ -61,6 +65,43 @@ function useProcessMedia({
...
+              alt,
+            },
+          };
+        },
+      });
+    },
+    [updateElementById]
+  );
+
+  const updateExistingElement = useCallback(
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js]:
```suggestion
  const updateExistingElementById = useCallback(
```

File: packages/story-editor/src/app/media/utils/useProcessMedia.js [hunk 9]
```diff
@@ -61,6 +65,43 @@ function useProcessMedia({
...
+              ...resource,
+            },
+          };
+        },
+      });
+    },
+    [updateElementById]
+  );
+
   const updateExistingElements = useCallback(
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/media/utils/useProcessMedia.js]:
```suggestion
  const updateExistingElementsByResourceId = useCallback(
```

**Comment by spacedmonkey** [comment]
I have add some recommendations on naming. https://github.com/GoogleForCreators/web-stories-wp/pull/10554#discussion_r809282631 https://github.com/GoogleForCreators/web-stories-wp/pull/10554#discussion_r809282991

**Comment by timarney** [comment]
Let me know if there's any action items to update on this one or if we'll move to new tickets.

**Comment by spacedmonkey** [comment]
Other than maybe renaming those functions, I think we are good to go.

**Comment by timarney** [comment]
> Other than maybe renaming those functions, I think we are good to go.

Updated

**Review by swissspidy — APPROVED** [review]

**Review by spacedmonkey — APPROVED** [review]

**Comment by swissspidy** [comment]
Follow-up: #10641
